stl/inc/ranges

```diff
@@ -1433,6 +1463,94 @@ namespace ranges {
         inline constexpr _Istream_fn<_Ty> istream;
     } // namespace views
 
+#if _HAS_CXX23
+    namespace _Pipe {
+        template <class _Ty>
+        void _Derived_from_range_adaptor_closure_impl(const volatile range_adaptor_closure<_Ty>&);
+
+        template <class _Ty>
+        concept _Derived_from_range_adaptor_closure = requires(const volatile _Ty& __obj) {
+            _Derived_from_range_adaptor_closure_impl(__obj);
+        };
+
+        template <class _Ty>
+        concept _Is_range_adaptor_common =
+            derived_from<remove_cvref_t<_Ty>, range_adaptor_closure<remove_cvref_t<_Ty>>> //
+            && _Derived_from_range_adaptor_closure<_Ty> && !range<_Ty>;
+
+        template <class _Rng, class _Ty>
+        concept _Is_invocable_range_adaptor_closure =
+            range<_Rng> && _Can_pipe<_Rng, _Ty> && _Is_range_adaptor_common<_Ty>;
+
+        template <class _Left, class _Right>
+        concept _Are_composable_range_adaptor_closures =
+            _Is_range_adaptor_common<_Left> && _Is_range_adaptor_common<_Right> && _Can_compose<_Left, _Right>;
+    } // namespace _Pipe
+
+    template <class _Derived>
+        requires is_class_v<_Derived> && same_as<_Derived, remove_cv_t<_Derived>>
+    class range_adaptor_closure {
+    public:
+        template <class _Other>
+            requires _Pipe::_Are_composable_range_adaptor_closures<_Derived, _Other>
+        constexpr auto operator|(_Other&& _Right) && noexcept(

```

Comment:
The fact that this does not have _all_ of the cvref qualifiers makes me uncomfortable, and I am wondering whether the standardese does require them or not.

Reply:
This made sense for the "old" implementation that only needed to support the well-behaved standard range adaptors, but the "new" implementation needs to support badly-behaved program-defined range adaptors because the wording allows them.